components/style/custom_properties.rs

```diff
@@ -97,52 +98,128 @@ impl ToCss for ComputedValue {
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct CustomPropertiesMap {
+pub struct OrderedMap<K, V>
+where
+    K: Eq + Hash + Clone,
+{
     /// Custom property name index.
-    index: Vec<Name>,
+    index: Vec<K>,
     /// Computed values indexed by custom property name.
-    values: HashMap<Name, ComputedValue>,
+    values: HashMap<K, V>,
 }
 
-impl CustomPropertiesMap {
+impl<K, V> OrderedMap<K, V>
+where
+    K: Eq + Hash + Clone,
+{
     /// Creates a new custom properties map.
     pub fn new() -> Self {
-        CustomPropertiesMap {
+        OrderedMap {
             index: Vec::new(),
             values: HashMap::new(),
         }
     }
 
-    /// Insert a computed value if it has not previously been inserted.
-    pub fn insert(&mut self, name: &Name, value: ComputedValue) {
-        debug_assert!(!self.index.contains(name));
-        self.index.push(name.clone());
-        self.values.insert(name.clone(), value);
+    /// Insert a custom property value.
+    pub fn insert(&mut self, name: K, value: V) {
+        if !self.values.contains_key(&name) {
+            self.index.push(name.clone());
+        }
+        self.values.insert(name, value);
     }
 
-    /// Custom property computed value getter by name.
-    pub fn get_computed_value(&self, name: &Name) -> Option<&ComputedValue> {
+    /// Custom property value getter by name.
+    pub fn get(&self, name: &K) -> Option<&V> {
         let value = self.values.get(name);
         debug_assert_eq!(value.is_some(), self.index.contains(name));
         value
     }
 
     /// Get the name of a custom property given its list index.
-    pub fn get_name_at(&self, index: u32) -> Option<&Name> {
+    pub fn get_key_at(&self, index: u32) -> Option<&K> {
         self.index.get(index as usize)
     }
 
-    /// Get an iterator for custom properties computed values.
-    pub fn iter(&self) -> hash_map::Iter<Name, ComputedValue> {
-        self.values.iter()
+    /// Get an iterator for custom properties values.
+    pub fn iter<'a>(&'a self) -> OrderedMapIterator<'a, K, V> {
+        OrderedMapIterator {
+            inner: self,
+            pos: 0,
+        }
     }
 
-    /// Get the count of custom properties computed values.
+    /// Get the count of custom properties values.
     pub fn len(&self) -> usize {
         debug_assert_eq!(self.values.len(), self.index.len());
         self.values.len()
     }
+
+    fn remove<Q: ?Sized>(&mut self, key: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        let index = match self.index.iter().position(|k| k.borrow() == key) {
+            Some(p) => p,
+            None => return None,
+        };
+        self.index.remove(index);
+        self.values.remove(key)
+    }
+}
+
+trait Iterable<K, V> {
+    fn index(&self) -> &Vec<K>;
+    fn values(&self) -> &HashMap<K, V>;
+}
+
+impl<K, V> Iterable<K, V> for OrderedMap<K, V>
+where
+    K: Eq + Hash + Clone,
+{
+    fn index(&self) -> &Vec<K> {
```

Comment:
Why is this trait needed? In Rust fields are private per-module, so you can use direct member access down below instead I think.